File: Python/Codes/Practice/LeetCodeCN/832 Flipping an image.py

class Solution:
    def flipAndInvertImage(self, A):
        """
        :type A: List[List[int]]
        :rtype: List[List[int]]
        """
        for v in A:
            size = len(v)
            for i in range(int(size/2)):
                v[i],v[size-1-i] = v[size-1-i] ^ 1, v[i] ^ 1
            
            # 奇数长度时，反转中位元素
            if (size % 2 != 0):
                v[int(size/2)] = v[int(size/2)] ^ 1
                
        return A


if __name__ == '__main__':
    s = Solution()
    result = s.flipAndInvertImage([[1,1,0],[1,0,1],[0,0,0]])
    print(result)
    result = s.flipAndInvertImage([[1,1,0,0],[1,0,0,1],[0,1,1,1],[1,0,1,0]])
    print(result)

File: Cpp/Codes/Practice/LeetCodeCN/832 Flipping an image.cpp
#include <gtest/gtest.h>
using namespace std;

vector<vector<int>> flipAndInvertImage(vector<vector<int>>& A) 
{
	for (auto &v : A)
	{
		size_t length = v.size();
		for (size_t i = 0; i < length / 2; ++i)
		{
			int t = v[length - 1 - i];
			v[length - 1 - i] = v[i] ^ 1;
			v[i] = t ^ 1;
		}

		if (length % 2 != 0)
		{
			v[length / 2] ^= 1;
		}
	}

	return A;
}

TEST(LeetCodeCN, tFlipAndInvertImage)
{
	vector<vector<int>> val1 = { {1, 1, 0},{1, 0, 1},{0, 0, 0} };
	auto result = flipAndInvertImage(val1);

	vector<vector<int>> val2 = { {1, 1, 0, 0},{1, 0, 0, 1},{0, 1, 1, 1},{1, 0, 1, 0} };
	result = flipAndInvertImage(val2);
}
